src/components/include/cbuf.h

```diff
@@ -405,241 +341,131 @@ __cbuf_alloc_meta_inconsistent(struct cbuf_alloc_desc *d, struct cbuf_meta *m)
-		/* 
-		 * This is complicated.
-		 *
-		 * See cbuf_slab_free for the rest of the story.  
-		 *
-		 * Assumptions: 
-		 * 
-		 * 1) The cbuf manager shared the cbuf meta (in the
-		 * meta_cbuf vector) information with this component.
-		 * It can remove asynchronously a cbuf from this
-		 * structure at any time IFF that cbuf is marked as
-		 * ~CBUF_IN_USE.
-		 *
-		 * 2) The slab descriptors, and the slab_desc vector
-		 * are _not_ shared with the cbuf manager for
-		 * complexity reasons.
-		 *
-		 * Question: How do we reconcile the fact that the
-		 * cbuf mgr might remove at any point a cbuf from this
-		 * component, but we still have a slab descriptor
-		 * lying around for it?  How will we know that the
-		 * cbuf has been removed, and not to use the slab
-		 * data-structure anymore?
-		 *
-		 * Answer: The slabs are deallocated lazily (seen
-		 * here).  When a slab is pulled off of the freelist
-		 * (see the following code), we check to make sure
-		 * that the cbuf meta information matches up with the
-		 * slab's information (i.e. the cbuf id and the
-		 * address in memory of the cbuf.  If they do not,
-		 * then we know that the slab is outdated and that the
-		 * cbuf backing it has been taken from this component.
-		 * In that case (shown in the following code), we
-		 * delete the slab descriptor.  Again, see
-		 * cbuf_slab_free to see the surprising fact that we
-		 * do _not_ deallocate the slab descriptor there to
-		 * reinforce that point.
-		 */
-		__cbuf_desc_free(d);
-		if (likely(!already_used)) {
-			cm->nfo.c.flags &= ~CBUFM_TOUCHED;
-			cm->nfo.c.refcnt--;
-		}
+	inconsistent = CBUF_INCONSISENT(cm);
+	if (unlikely(inconsistent)) {
+		CBUFM_SET_NEXT_NULL(cm);
 		goto again;
 	}
-
-	if (tmem) {
+	unsigned int p, n;
+	p = cm->nfo;
+	n = p+CBUFM_INC_UNIT;
+	if (unlikely(!cos_cas((unsigned long *)&cm->nfo, p, n))) {
+		CBUFM_SET_NEXT_NULL(cm);
+		goto again;
+	}
+	CBUFM_SET_NEXT_NULL(cm);
+	cm->owner_nfo.c.nsent = cm->owner_nfo.c.nrecvd = 0;
+	/*if (unlikely(tmem)) {
 		cm->owner_nfo.thdid = cos_get_thd_id();
 		cos_comp_info.cos_tmem_available[COMP_INFO_TMEM_CBUF]--;
 		assert(cm->nfo.c.flags & CBUFM_TMEM);
-	} else {
-		cm->owner_nfo.c.nsent = cm->owner_nfo.c.nrecvd = 0;		
-	}
-	ret = (void*)(cm->nfo.c.ptr << PAGE_ORDER);
+	} */     //???
 done:
-	*cb = cbuf_cons(cbid, len, tmem);
-	CBUF_RELEASE();
+	if (unlikely(tmem)) CBUF_SET_TMEM(cm);
+	ret = (void *)(CBUFM_GET_PTR(cm) << PAGE_ORDER);
+	cbid = cm->cbid.cbid;
+	*cb = cbuf_cons(cbid);
 	return ret;
 }
 
-static inline void *
-cbuf_alloc(unsigned int sz, cbuf_t *cb) { return __cbuf_alloc(sz, cb, 1); }
-static inline void *
-cbufp_alloc(unsigned int sz, cbufp_t *cb)  { return __cbuf_alloc(sz, (cbuf_t*)cb, 0); }
-
 /* 
  * precondition:  cbuf lock must be taken.
  * postcondition: cbuf lock has been released.
  */
 static inline void
-__cbuf_done(int cbid, int tmem)
+cbuf_free(cbuf_t cb)
```

Comment:
I'm not sure it is reasonable to expect the user to always track the whole cb struct, and pass it in to free.  Perhaps it is, but I think that components like the file-system (when reprogrammed to use cbufs) will be the test.